### PR TITLE
chore: Get All Destinations API ADR

### PR DESCRIPTION
**Pull request by tomfrenken**
This ticket concludes our ADR discussion on the Get All Destinations API ADR.

Part of SAP/cloud-sdk-backlog#738.

### CODE REVIEW

**Review by deekshas8 — COMMENTED** [review]
LGTM.

File: knowledge-base/adr/0034-get-destinations.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+export async function getAllDestinationsFromDestinationService(
+  options: DestinationOptions
+): Promise<Destination[] | null>
+```
+
+This function should throw a warning if one of the destinations is incomplete/compromised.
+
+If our users have feature requests for the other look-up functions, e.g. `searchEnvVariablesForDestination` etc., we shall implement them on-demand.
+
+If we end up implementing every look-up function to get all destinations, we are to implement the parent function and deprecate and remove the predecessors functions.
```
Comment (FrankEssenberger) [inline knowledge-base/adr/0034-get-destinations.md:43]: [req] Please add also the options for fetching and the Return type to the ADR.

File: .changeset/unlucky-cooks-work.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@sap-cloud-sdk/connectivity': minor
+---
+
+[New Functionality] Support fetching all destinations from the destination service simultaneously.
```
Comment (FrankEssenberger) [inline .changeset/unlucky-cooks-work.md:5]: [req] Be a bit more precise like: All destinations of a subacction and service instance simultaneously.

File: packages/connectivity/src/scp-cf/destination/destination-accessor-provider-subscriber-lookup.spec.ts [hunk 3]
```diff
@@ -296,3 +331,63 @@ describe('jwtType x selection strategy combinations. Possible values are {subscr
...
+      messageContext: 'destination-accessor'
+    });
+    const warnSpy = jest.spyOn(logger, 'warn');
+
+    const allDestinations = await getAllDestinationsFromDestinationService();
+    expect(allDestinations).toEqual(null);
+    expect(warnSpy).toHaveBeenCalledWith(
+      'Could not retrieve destinations from destination service.'
+    );
+  });
```
Comment (FrankEssenberger) [inline packages/connectivity/src/scp-cf/destination/destination-accessor-provider-subscriber-lookup.spec.ts:392]: [req] Please add a test where the response of the destination service is service empty. See the specs (empty array) 
https://api.sap.com/api/SAP_CP_CF_Connectivity_Destination/resource. 

Also an error case like 404 or 500 would be nice to mock to ensure we give a resonable exception.

File: packages/connectivity/src/scp-cf/destination/destination-accessor.ts [hunk 4]
```diff
@@ -85,3 +103,89 @@ export async function getDestination(
...
+      token.decoded.iss!
+    )}" from destination service.`
+  );
+
+  const [instance, subaccount] = await Promise.all([
+    fetchInstanceDestinations(destinationServiceUri, token.encoded, options),
+    fetchSubaccountDestinations(destinationServiceUri, token.encoded, options)
+  ]);
+
+  let loggerMessage = '';
```
Comment (FrankEssenberger) [inline packages/connectivity/src/scp-cf/destination/destination-accessor.ts:160]: [pp] create helper methods for the logging

File: packages/connectivity/src/scp-cf/destination/destination-accessor.ts [hunk 5]
```diff
@@ -85,3 +103,89 @@ export async function getDestination(
...
+  const allDestinations = instance.concat(subaccount);
+
+  if (allDestinations && allDestinations.length !== 0) {
+    logger.debug(
+      `Successfully retrieved all destinations for account: "${parseSubdomain(
+        token.decoded.iss!
+      )}" from destination service.`
+    );
+  } else {
+    logger.warn('Could not retrieve destinations from destination service.');
```
Comment (FrankEssenberger) [inline packages/connectivity/src/scp-cf/destination/destination-accessor.ts:180]: [req] If there are no destination this could be - I would not warn

Comment (FrankEssenberger) [reply packages/connectivity/src/scp-cf/destination/destination-accessor.ts:180]: [pp] I would prefer an empty array in a non error case an not `null`

**Review by FrankEssenberger — CHANGES_REQUESTED** [review]
Nothing big a few things requested. Overall looks good.

File: packages/connectivity/src/scp-cf/destination/destination-accessor-provider-subscriber-lookup.spec.ts (hunk 3 above)
Comment (tomfrenken) [reply packages/connectivity/src/scp-cf/destination/destination-accessor-provider-subscriber-lookup.spec.ts:392]: I think with the `mockThingsForCombinations(true, true);` I already test the empty case, however, I guess I can add some 404/500 case.

File: packages/connectivity/src/scp-cf/destination/destination-accessor.ts [hunk 6]
```diff
@@ -85,3 +102,79 @@ export async function getDestination(
...
   return destination;
 }
+
+/**
+ * Creates comprehensive log messages from a destinations array and their origin.
+ * @param origin - Origin of the destination.
+ * @param destinations - Array of destinations.
+ * @returns Logs of the retrival of destinations.
+ */
+function destinationFetchLogHelper(
```
Comment (FrankEssenberger) [inline packages/connectivity/src/scp-cf/destination/destination-accessor.ts:112]: [pp] the `helper`feels a bit strange and our style guide also says: user verbs (https://github.com/SAP/cloud-sdk-js/blob/main/STYLEGUIDE.md).

File: packages/connectivity/src/scp-cf/destination/destination-accessor-provider-subscriber-lookup.spec.ts [hunk 7]
```diff
@@ -296,3 +350,83 @@ describe('jwtType x selection strategy combinations. Possible values are {subscr
...
+
+    expect(debugSpy).toHaveBeenCalledWith(
+      'Retrieving all destinations for account: "subscriber" from destination service.'
+    );
+    expect(debugSpy).toHaveBeenCalledWith(
+      'Retrieving subaccount destination: DESTINATION.\n'
+    );
+  });
+
+  it('should fetch all provider destinations when called without anything', async () => {
```
Comment (FrankEssenberger) [inline packages/connectivity/src/scp-cf/destination/destination-accessor-provider-subscriber-lookup.spec.ts:390]: [pp] without anything is perhaps a misleading doupble negation `without passing a JWT` perhaps?

File: packages/connectivity/src/scp-cf/destination/destination-accessor-provider-subscriber-lookup.spec.ts [hunk 8]
```diff
@@ -296,3 +350,83 @@ describe('jwtType x selection strategy combinations. Possible values are {subscr
...
+    const logger = createLogger({
+      package: 'connectivity',
+      messageContext: 'destination-accessor'
+    });
+    const debugSpy = jest.spyOn(logger, 'debug');
+
+    const allDestinations = await getAllDestinationsFromDestinationService();
+    expect(allDestinations).toEqual([]);
+    expect(debugSpy).toHaveBeenCalledWith(
+      'Could not retrieve destinations from destination service.'
```
Comment (FrankEssenberger) [inline packages/connectivity/src/scp-cf/destination/destination-accessor-provider-subscriber-lookup.spec.ts:409]: [pp] This log statement is perhaps missleading if thre are simply no destinaitons.

File: packages/connectivity/src/scp-cf/destination/destination-from-service.ts [hunk 9]
```diff
@@ -264,6 +271,20 @@ class DestinationFromServiceRetriever {
     return { encoded, decoded: decodeJwt(encoded) };
   }
 
+  private static checkDestinationForCustomJwt(destination: Destination): void {
```
Comment (FrankEssenberger) [inline packages/connectivity/src/scp-cf/destination/destination-from-service.ts:274]: [q] This is marked as a change here, but I guess nothing changed for the two private static methods?

**Review by FrankEssenberger — APPROVED** [review]
LGTM- only a few minor questions on naming of the tests and content of log statements

File: packages/connectivity/src/scp-cf/destination/destination-from-service.ts (hunk 9 above)
Comment (tomfrenken) [reply packages/connectivity/src/scp-cf/destination/destination-from-service.ts:274]: Yes, they were just moved because private has to come before public
